dbr quant: split observer insertion to a separate pass

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71551 dbr quant: record dag of non-quantizeable ops
* #71324 dbr quant: refactor first_call hooks to be standalone
* #71312 dbr quant: rename seen_op_info to seen_q_op_info
* #71267 dbr quant: make SeenOpInfo a dataclass
* **#71253 dbr quant: split observer insertion to a separate pass**
* #71245 dbr quant: make QTensorInfo a dataclass and add orig_dtype
* #71233 dbr quant: do not crash on unsupported qconfig_dict keys if they are empty

Summary:

Before this PR, observers were inserted at the same time as
we recorded ops seen while tracing with example input. This is not
ideal because for function fusion (not yet implemented),
we need to be able to look ahead from the current op to properly
insert observers.

This PR refactors observer insertion in DBR quantization to happen
in a separate pass after the ops are recorded.  There is no functionality
change in this diff, but this PR will make it easier to implement
function fusion in a future PR.

Note: the qconfig is still used during tracing to assign each
op an inference dtype. This is not ideal, in the future we may move this
step to happen as a separate pass as well. The reason we keep it as is
in this PR because some more refactoring would be needed to allow
this to both happen in a separate pass as well as survive module
boundaries.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33558280](https://our.internmc.facebook.com/intern/diff/D33558280)

vkuzo committed:
dbr quant: split observer insertion to a separate pass

Summary:

Before this PR, observers were inserted at the same time as
we recorded ops seen while tracing with example input. This is not
ideal because for function fusion (not yet implemented),
we need to be able to look ahead from the current op to properly
insert observers.

This PR refactors observer insertion in DBR quantization to happen
in a separate pass after the ops are recorded.  There is no functionality
change in this diff, but this PR will make it easier to implement
function fusion in a future PR.

Note: the qconfig is still used during tracing to assign each
op an inference dtype. This is not ideal, in the future we may move this
step to happen as a separate pass as well. The reason we keep it as is
in this PR because some more refactoring would be needed to allow
this to both happen in a separate pass as well as survive module
boundaries.

Test plan:

```
python test/test_quantization.py -k DBR
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33558280).

vkuzo committed:
Update on "dbr quant: split observer insertion to a separate pass"


Summary:

Before this PR, observers were inserted at the same time as
we recorded ops seen while tracing with example input. This is not
ideal because for function fusion (not yet implemented),
we need to be able to look ahead from the current op to properly
insert observers.

This PR refactors observer insertion in DBR quantization to happen
in a separate pass after the ops are recorded.  There is no functionality
change in this diff, but this PR will make it easier to implement
function fusion in a future PR.

Note: the qconfig is still used during tracing to assign each
op an inference dtype. This is not ideal, in the future we may move this
step to happen as a separate pass as well. The reason we keep it as is
in this PR because some more refactoring would be needed to allow
this to both happen in a separate pass as well as survive module
boundaries.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33558280](https://our.internmc.facebook.com/intern/diff/D33558280)

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/_dbr/auto_trace.py:
```diff
@@ -325,6 +323,12 @@ def _patched_module_call(self, *args, **kwargs):
+                if first_call:
+                    for _, v in self.named_modules():
+                        if hasattr(v, '_auto_quant_state'):
+                            v._auto_quant_state.insert_observers(v)
```
Just so I understand, this is functionally the same as before, where we inserted the observers in `_first_call_op_prepare_before_hook_create_subgraphs`, because here we're calling `insert_observers` after executing the op in L325?

vkuzo replied:
Correct.  Here is a diagram of before and after:

```
# op graph
x0 -> op0 -> op1 -> x1

# before:
1. trace through op0, insert observers for op0
2. trace through op1, insert observers for op1

# after:
1. trace through op0 and op1
2. insert observers for op0 and op1
```

andrewor14 commented on torch/ao/quantization/_dbr/utils.py:
```diff
@@ -281,27 +281,27 @@ class FuncOutputDTypeType(enum.Enum):
-    op: Callable,
-    args: Tuple[Any, ...],
-    op_packing_only_uses_module_attributes: bool,
+    seen_op_info: SeenOpInfo,
```
This part is just clean up right? (not strictly needed for function fusion?)

vkuzo replied:
correct, this is a refactor (after code was moved around, it was easier to just pass `seen_op_info` here which contains all the metadata instead of individual metadata instances)

vkuzo committed:
Update on "dbr quant: split observer insertion to a separate pass"


Summary:

Before this PR, observers were inserted at the same time as
we recorded ops seen while tracing with example input. This is not
ideal because for function fusion (not yet implemented),
we need to be able to look ahead from the current op to properly
insert observers.

This PR refactors observer insertion in DBR quantization to happen
in a separate pass after the ops are recorded.  There is no functionality
change in this diff, but this PR will make it easier to implement
function fusion in a future PR.

Note: the qconfig is still used during tracing to assign each
op an inference dtype. This is not ideal, in the future we may move this
step to happen as a separate pass as well. The reason we keep it as is
in this PR because some more refactoring would be needed to allow
this to both happen in a separate pass as well as survive module
boundaries.

Test plan:

```
python test/test_quantization.py -k DBR
```

Differential Revision: [D33558280](https://our.internmc.facebook.com/intern/diff/D33558280)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33558280).
